exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/project/ProjectMemoryManager.java

```diff
@@ -0,0 +1,323 @@
+            String columnName = vv.getField().getName();
+            this.name = columnName;
+        }
+
+        public OutputWidthExpression getOutputExpression() { return outputExpression; }
+
+        public OutputColumnType getOutputColumnType() { return outputColumnType; }
+
+        boolean isFixedWidth() { return widthType == WidthType.FIXED; }
+
+        public int getWidth() { return width; }
+
+        public String getName() { return name; }
+    }
+
+    void ShoulgNotReachHere() {
+        throw new IllegalStateException();
+    }
+
+    private void setIncomingBatch(RecordBatch recordBatch) {
+        incomingBatch = recordBatch;
+    }
+
+    private void setOutgoingBatch(ProjectRecordBatch outgoingBatch) {
+        this.outgoingBatch = outgoingBatch;
+    }
+
+    public ProjectMemoryManager(int configuredOutputSize) {
+        super(configuredOutputSize);
+        outputColumnSizes = new HashMap<>();
+    }
+
+    public boolean isComplex(MajorType majorType) {
+        MinorType minorType = majorType.getMinorType();
+        return minorType == MinorType.MAP || minorType == MinorType.UNION || minorType == MinorType.LIST;
+    }
+
+    boolean isFixedWidth(TypedFieldId fieldId) {
+        ValueVector vv = getOutgoingValueVector(fieldId);
+        return (vv instanceof FixedWidthVector);
+    }
+
+    public ValueVector getOutgoingValueVector(TypedFieldId fieldId) {
+        Class<?> clazz = fieldId.getIntermediateClass();
+        int[] fieldIds = fieldId.getFieldIds();
+        return outgoingBatch.getValueAccessorById(clazz, fieldIds).getValueVector();
+    }
+
+    static boolean isFixedWidth(ValueVector vv) {  return (vv instanceof FixedWidthVector); }
+
+    static final int BIT_VECTOR_WIDTH = UInt1Vector.VALUE_WIDTH;
+    static final int OFFSET_VECTOR_WIDTH = UInt4Vector.VALUE_WIDTH;
+
+    static int getWidthOfFixedWidthType(MajorType majorType) {
+        DataMode mode = majorType.getMode();
+        MinorType minorType = majorType.getMinorType();
+        final boolean isVariableWidth  = (minorType == MinorType.VARCHAR || minorType == MinorType.VAR16CHAR
+                || minorType == MinorType.VARBINARY);
+
+
+        if (isVariableWidth) {
+            throw new IllegalArgumentException("getWidthOfFixedWidthType() cannot handle variable width types");
+        }
+
+        final boolean isOptional = (mode == DataMode.OPTIONAL);
+        final boolean isRepeated = (mode == DataMode.REPEATED);
+
+        int stdDataSize = 0;
+
+        if (isOptional) { stdDataSize += BIT_VECTOR_WIDTH; }
+
+        if (minorType != MinorType.NULL) {
+            stdDataSize += TypeHelper.getSize(majorType);
+        }
+
+        if (isRepeated) {
+            stdDataSize = (stdDataSize * STD_REPETITION_FACTOR) + OFFSET_VECTOR_WIDTH;
+        }
+
+        return stdDataSize;
+    }
+
+    static int getWidthOfFixedWidthType(ValueVector vv) {
+        assert isFixedWidth(vv);
+        return vv.getPayloadByteCount(1);
```

Comment:
instead of getPayloadByteCount, isn't there a function like getValueWidth or something like that ?